### ISSUE TITLE
Add to configuration hardcoded timeout of HttpClient

### ISSUE THREAD

**Issue by RightDecision2** [ThreeMammals/Ocelot#1869]
## Expected Behavior / New Feature
// HttpClientBuilder.cs
`_defaultTimeout = TimeSpan.FromSeconds(_configuration.TimeOut);`

## Actual Behavior / Motivation for New Feature
// HttpClientBuilder.cs
`_defaultTimeout = TimeSpan.FromSeconds(90);`

## Specifications

  - Version: All
  - Platform: All
  - Subsystem: Any

**Comment by raman-m** [comment]
Are you going to create a PR these days?

**Comment by RightDecision2** [comment]
Yes

вт, 26 дек. 2023 г., 22:40 Raman Maksimchuk ***@***.***>:

**Comment by raman-m** [comment]
What is the problem with default 90 seconds timeout?

As a quick fix use QoS [TimeoutValue](https://ocelot.readthedocs.io/en/latest/features/qualityofservice.html#quality-of-service:~:text=You%20can%20set%20the%20TimeoutValue%20in%20isolation%20of%20the%20ExceptionsAllowedBeforeBreaking%20and%20DurationOfBreak%20options%3A) property to have custom timeout value. 
I don't think that performance will decrease dramatically... A few microseconds will be added 😄

**Comment by raman-m** [comment]
>@RightDecision2 commented [on Dec 26](https://github.com/ThreeMammals/Ocelot/issues/1869#issuecomment-1869713955)

👌 Great! Waiting for your PR creation... Will review with pleasure!

**Comment by RightDecision2** [comment]
I have OCR service, which process pdf file. Every request average 2mins.
When I use QuOS its random cancelled after 90 seconds (few times might be
worked).
So fo me, is only one possible solution - increase default timeout. I think
for someone it could be usefull to.

вт, 26 дек. 2023 г. в 22:52, Raman Maksimchuk ***@***.***>:

**Comment by raman-m** [comment]
Dear author, What's your name?
What's your LinkedIn?

**Comment by raman-m** [comment]
> I have OCR service, which process pdf file. Every request average 2mins.
> **When I use QoS its random cancelled after 90 seconds (few times might be worked).**
> 

Strange cancellation of your requests by QoS logic... You need to show this behavior by showing us Ocelot logs.
It should not be cancelled randomly... All requests should be done without problems if `TimeoutValue` in QoS is greater than 2 mins (average processing time by your downstream service). If some requests failed that means root cause is not on Ocelot's side...

What is the maximum processing time? Seems you have to define max value in route config.
**What exact value of TimeoutValue?**
### Please attach route JSON!

**Comment by raman-m** [comment]
>So for me, is only one possible solution - increase default timeout. I think for someone it could be useful to.

I'm not sure it will help you...
How stable your downstream service if you connect directly (without Ocelot routing)?

As we've agreed on, you are going to create a PR soon to manage default timeout from route config, right?

**Comment by RightDecision2** [comment]
Yes, when connected directly all working fine (~2minutes for every request).
Yes, would be great to manage default timeouts from config file ocelot.json

ср, 27 дек. 2023 г. в 13:03, Raman Maksimchuk ***@***.***>:

**Comment by lemosluan** [comment]
The doc from ocelot's site says that is a timeout to 90s.
You can use Polly to reach a new timeout, but in some cases, the response from our services take more than 90s, so it would be useful if the timeout on startup of ocelot can be set.

![image](https://github.com/ThreeMammals/Ocelot/assets/34046499/1e6055d4-bf8d-4183-ab51-1d909434b635)

**Comment by raman-m** [comment]
@RightDecision2 commented [on Jan 9](https://github.com/ThreeMammals/Ocelot/issues/1869#issuecomment-1882501826)

How is your development going?
I guess you should be interested in fixing your problem...
But sorry, as a team, we are busy with other more important tasks. So, you need to fix by your own and create a PR, if you wish to contribute.

**Will you contribute or not?**

Можешь писать на русском языке, если это ускорит нашу разработку.
Не люблю, когда скрываются под всякими анонимными никами, да ещё и отвечают с мобилы, где установлен русскоязычный почтовый клиент. 
И я задавал этот [вопрос](https://github.com/ThreeMammals/Ocelot/issues/1869#issuecomment-1870085487)
Зачем шифроваться под левой учёткой ГитХаба? Не понимаю этого! Наверняка есть основной аккаунт, другой...

**Comment by raman-m** [comment]
> @lemosluan commented [on Feb 8](https://github.com/ThreeMammals/Ocelot/issues/1869#issuecomment-1934279984)

Hi Luan!
What's your solution for Ocelot pipeline setup?

This author's problem can be fixed by a [Delegating Handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html).
A few lines of code in the body of the handler. Something like that:
```csharp
public class TimeoutHandler : DelegatingHandler
{
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken token)
    {
            TimeSpan? timeout = TimeSpan.FromHours(1.0); // hardcoded but can be read from the config
            request.Properties["RequestTimeout"] = timeout;
            return await base.SendAsync(request, token);
    }
}
```
But I'm not sure on design correctness, it should be checked.
Personally I like the following robust solution: [Better timeout handling with HttpClient](https://thomaslevesque.com/2018/02/25/better-timeout-handling-with-httpclient/)

And after that, just attach `TimeoutHandler` to Ocelot pipeline as:
```c#
ConfigureServices(s => s
    .AddOcelot()
    .AddDelegatingHandler<TimeoutHandler>()
)
```
and route config is
```json
{
  // other props
  "DelegatingHandlers": ["TimeoutHandler"]
}
```

**Comment by RightDecision2** [comment]
Роман, здравствуйте. Учетки не левые, создавал много лет назад. Почта
основная.
Касательно разработки, прогресса нет. Всё вызовы к службе OCR пустил пока
напрямую, минуя Ocelot. Служба читает штрихкоды с файлов pdf, иногда
занимает много времени.
Я видел, что в настройках можно использовать qos и timeout будет браться из
конфигурации, по исходниками тоже на это похоже. Но на проде все равно
иногда проскакивал timeout к службе OCR, пришлось сделать вызовы напрямую к
службам. Теперь timeout не вылетает, но в Ocelot имеется дополнительный
middleware и Consul Service Discovery.

Я вас понял. Вопрос не горит, но было бы здорово иметь такую возможность,
настройки timeout, через файл конфигурации.

чт, 22 февр. 2024 г., 22:47 Raman Maksimchuk ***@***.***>:

**Comment by raman-m** [comment]
@RightDecision2 Приоритет [низкий](https://github.com/ThreeMammals/Ocelot/issues/1869#event-11411847086)!
Без понятия в какой релиз это попадёт. Если кто-то создаст ПР, то фича появится быстрее.
Ждите! Может кто-то контрибьютнет...

**Comment by hogwartsdeveloper** [comment]
I would like to do it.

**Comment by raman-m** [comment]
> I would like to do it.

@hogwartsdeveloper Dear Zhannur,
You are assigned!
